site/website/blog/2020-10-06-enabling-full-text-search.md

```diff
@@ -0,0 +1,492 @@
+`mapStateful` step. This can be performed either globally or partitioned
+on a grouping key. If all your records share a common key it is always a
+good idea to do the grouping by the key, because the mapping state is
+then evenly distributed across all nodes and the mapping operation
+parallelized.
+
+Our visit record contains the pet id but not the owner id, in such case,
+so we don’t have a single grouping key for all records. There are two options: 
+* perform the stateful mapping globally, which has the advantage
+of having a single state - the joining logic is simpler and the pipeline
+more straightforward. The obvious disadvantage is that such a solution
+is not scalable because the state may not fit onto a single member or it
+cannot be performed by a single thread;
+
+* create two separate mapping steps - one grouped on the owner id and
+the other on the pet id. This is a trade-off between simplicity and
+scalability.
+
+There are several observations which hold for both implementations:
+
+* the mapping state must cover all data from the beginning, there might
+be a domain-specific rule that would allow eviction of old items;
+
+* the records might arrive in any order, even if there is an FK between
+the records in the database, the order is not guaranteed, this is a
+property of CDC.
+
+* once the object is emitted from the join step it must not be modified,
+otherwise the later stages might see it in inconsistent state, causing
+issues which are hard to debug. We make defensive copies to avoid this.
+
+Go to
+[Github](https://github.com/hazelcast-demos/pet-clinic-index-job/blob/master/src/main/java/org/example/jet/petclinic/JoiningState.java#L11)
+to see the full source code.
+
+### Deleting changes
+
+The `ChangeRecord#operation()` method provides information about what
+kind of operation was performed on the record. If you need to handle
+deletions you should use it. The Petclinic application doesn’t allow
+delete operations, so we don’t handle those.
+
+### Elastic sink
+
+The fourth key part of the pipeline is the Elastic sink. It takes a
+document and writes it to Elasticsearch. You as a developer need to
+provide information where the Elastic instance is running, what index to
+write to and how to convert the `Document` into one of `IndexRequest`,
+`UpdateRequest` or `DeleteRequest`. The Sink then takes care of batching
+the requests and parallelizing for higher performance, retries in case
+of network issues.
+
+### The Pipeline
+
+The final pipeline composes all the individual steps in a
+straightforward way:
+
+```java
+Pipeline p = Pipeline.create();
+p.readFrom(mysqlSource)
+ .withoutTimestamps()
+ .map(PetClinicIndexJob::mapChangeRecordToPOJO).setName("mapChangeRecordToPOJO")
+ .mapUsingService(keywordService, PetClinicIndexJob::enrichWithKeywords).setName("enrichWithKeywords")
+ .mapStateful(JoiningState::new, JoiningState::join).setName("JoiningState::join")
+ .writeTo(elasticSink);
+```
+
+Or visualized as a graph:
```

Comment:
Do these graph visualisations really add value? Maybe remove them?

Reply:
I think so, especially the second one, which is more complicated and I don't show the code here. Then I think it's good to see the first one visualised.